File: 3.math-stats/permutation/permuteAll.js
const permuteAll = function (nums) {
  let res = [];
  permute(nums, [], res);
  return res;
};
const permute = function (nums, path, res) {
  if (nums.length < 1) {
    res.push(path);
    return res;
  }
  for (let i = 0; i < nums.length; ++i) {
    const newNums = [...nums];
    const item = newNums.splice(i, 1);
    const newPath = [...path, ...item];
    permute(newNums, newPath, res);
  }
};

nums = [1, 2, 3];
console.log(permuteAll(nums));

File: 3.math-stats/permutation/permute_all.rb
def permute_all(nums)
  res = []
  permute(nums, [], res)
  return res
end

def permute(nums, path, res)
  if nums.nil? || nums.empty?
    res.push(path)
    return res
  end
  for i in 0..(nums.length - 1)
    newNums = nums.dup
    newNums.delete_at(i)
    permute(newNums, path + [nums[i]], res)
  end
end

nums = [1, 2, 3]
p(permute_all(nums))
